pkg/upgrade/upgrade.go

```diff
@@ -0,0 +1,246 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/mitchellh/go-ps"
+	"github.com/opencontainers/selinux/go-selinux"
+	"github.com/rancher/k3s-upgrade/pkg/types"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+func GetProcessPID() (int, error) {
+	processes, err := ps.Processes()
+	if err != nil {
+		return 0, err
+	}
+	for _, proc := range processes {
+		if strings.HasSuffix(proc.Executable(), "k3s") {
+			if strings.Contains(proc.Executable(), "init") || strings.Contains(proc.Executable(), "channelserver") {
+				continue
+			}
+			return proc.Pid(), nil
+		}
+	}
+	return -1, fmt.Errorf("k3s process is not running")
+}
+
+func GetBinPath(pid int) (string, error) {
+	_, err := os.FindProcess(pid)
+	if err != nil {
+		return "", fmt.Errorf(" process is not found")
+	}
+	if pid == 1 {
+		return "/bin/k3s", nil
+	}
+	binPath := path.Join("/host/proc", strconv.Itoa(pid), "cmdline")
+	if fileExists(binPath) {
+		return binPath, nil
+	}
+	return "", fmt.Errorf(" failed to fetch the k3s binary path from process %d", pid)
+}
+
+func TermProcess(pid int) error {
+	proc, err := os.FindProcess(pid)
+	if err != nil {
+		fmt.Errorf(" process is not found")
+	}
+	return proc.Signal(syscall.SIGTERM)
+}
+
+func MatchSelector(hostPath, configPath string, preservedEntries []string) error {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	conf, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
```

Comment:
so since this will run in a pod, the kubeconfig will not be actually supplied here, so it should use the secret token mounted within a pod which is called inclusterconfig, an example for that is:
```
 // creates the in-cluster config
    config, err := rest.InClusterConfig()
    if err != nil {
        panic(err.Error())
    }
    // creates the clientset
    clientset, err := kubernetes.NewForConfig(config)
```

Reply:
because the KUBECONFIG will be empty in the pod, the `clientcmd.BuildConfigFromFlags()` call devolves to `rest.InClusterConfig()` to establish the config and clientset from that